Draft: DIP 20 - Dash Opcode updates (#64)
dashpay/dips: dip-martmangus-newopcodes.md

<pre>
Title: Activating old and implementing new opcodes
Author: Mart Mangus
Comments: https://github.com/Dash-Dev-EE/DIP-draft/issues
Status: Draft
Layer: Consensus (hard fork)
Created: 2020-06-01
License: MIT License
</pre>

# Abstract

> This DIP describes reactivation of a disabled opcodes (`OP_CAT`, `OP_AND`, `OP_OR`, `OP_XOR`, `OP_DIV`, `OP_MOD`) and activation of new opcodes (`OP_SPLIT`, `OP_BIN2NUM`, `OP_NUM2BIN`, `OP_CHECKDATASIG` and `OP_CHECKDATASIGVERIFY`) to expand the use of Dash scripting system.

Reviewer: ```suggestion
This DIP describes reactivation of several disabled opcodes (`OP_CAT`, `OP_AND`, `OP_OR`, `OP_XOR`, `OP_DIV`, `OP_MOD`) and activation of several new opcodes (`OP_SPLIT`, `OP_BIN2NUM`, `OP_NUM2BIN`, `OP_CHECKDATASIG` and `OP_CHECKDATASIGVERIFY`) to expand the use of the Dash scripting system.
```

# Motivation

> Several opcodes were disabled in the Bitcoin scripting system due to discovery of series of bugs in early days of Bitcoin. The functionality of these opcodes has been re-examined by Bitcoin Cash developers few years ago. Many of the disabled opcodes have been enabled and few of them re-designed to replace the original ones. In addition, couple of new opcodes have been written and implemented to improve Bitcoin scripting system in Bitcoin Cash even further. Implementing these opcodes into Dash is necessary for broadening the functionality of the system and enabling developers to build new solutions, which in turn would expand the use of Dash.

Reviewer: ```suggestion
Several opcodes were disabled in the Bitcoin scripting system due to the discovery of a series of bugs in the early days of Bitcoin. The functionality of these opcodes was re-examined by Bitcoin Cash developers few years ago. Many of the disabled opcodes have been enabled and several of them re-designed to replace the original ones. In addition, several new opcodes have been written and implemented to further improve the Bitcoin scripting system in Bitcoin Cash. Implementing these opcodes into Dash is necessary for broadening the functionality of the system and enabling developers to build new solutions, which in turn would expand the use of Dash.
```

|OP_SPLIT              |127    |0x7f|x n   |x1 x2   |Split byte array *x* at position *n*                   |
|OP_NUM2BIN            |128    |0x80|a b   |out     |Convert numeric *a* into byte array of length *b*      |
|OP_BIN2NUM            |129    |0x81|x     |out     |Convert byte array *x* into numeric                    |
|OP_AND                |132    |0x84|x1 x2 |out     |Boolean *AND* between each bit of the inputs           |
|OP_OR                 |133    |0x85|x1 x2 |out     |Boolean *OR* between each bit of the inputs            |
|OP_XOR                |134    |0x86|x1 x2 |out     |Boolean *EXCLUSIVE OR* between each bit of the inputs  |
|OP_DIV                |150    |0x96|a b   |out     |*a* is divided by *b*                                  |
|OP_MOD                |151    |0x97|a b   |out     |return the remainder after *a* is divided by *b*       |
|OP_CHECKDATASIG       |186    |0xba|sig&nbsp;msg&nbsp;pk|out     |If signature *sig* is valid, output to the stack|
|OP_CHECKDATASIGVERIFY |187    |0xbb|sig&nbsp;msg&nbsp;pk|-       |If signature *sig* is valid, false will cause the script to fail|

## OP_CAT

    Opcode (decimal): 126
    Opcode (hex): 0x7e

> `OP_CAT` takes two byte arrays from the stack, concates and pushes the result back to the stack.

Reviewer: ```suggestion
`OP_CAT` takes two byte arrays from the stack, concatenates them and pushes the result back to the stack.
```

	x1 x2 OP_CAT → out

Example:
* `0x11 0x2233 OP_CAT → 0x112233`

References:
* [Detailed specification in Bitcoin Cash](https://github.com/bitcoincashorg/bitcoincash.org/blob/master/spec/may-2018-reenabled-opcodes.md#op_cat)
* [Implementation in Bitcoin ABC](https://reviews.bitcoinabc.org/D1097)

## OP_SPLIT

    Opcode (decimal): 127
    Opcode (hex): 0x7f
    

> `OP_SPLIT` is inverse of `OP_CAT` and a replacement operation for disabled opcodes `OP_SUBSTR`, `OP_LEFT` and `OP_RIGHT`.

Reviewer: ```suggestion
`OP_SPLIT` is the inverse of `OP_CAT` and a replacement operation for disabled opcodes `OP_SUBSTR`, `OP_LEFT` and `OP_RIGHT`.
```
